### PR TITLE
Disable Rate Limiting when editing Blocks in admin

### LINKED ISSUE #7738: Rate-Limiting blocking edits to admin page blocks
## Describe the bug

Due to daily abuse of our StoreApi checkout endpoint, we were thrilled to see the new rate-limiting feature released in the Woo Blocks feature plugin. We'd already tried many other various ways of preventing hundreds of failed orders via credit card fraud attempts. The StoreApi endpoint makes abuse trivial, so rate-limiting is a welcome, and necessary, feature.

However, because the same blocks and endpoints are used in the admin editor, we're now also getting blocked from saving our updates to pages. Screenshot below. If you (or a team on the same corporate static IP address) make too many changes in a small window, editing pages/posts with blocks becomes unusable.

## To reproduce

1. Go to any admin page and create a product-related block or blocks.
2. Make many changes to the block(s) and keep saving them.
3. Shortly, you will get a notice that you've been rate-limited.

## Expected behavior

Any logged-in administrator should be excluded from the rate limit. Since both IP and UserID are known, it should be pretty straightforward to exclude any UserIDs with the role of `Administrator`.

## Screenshots

image

### WordPress (please complete the following information):

-   WordPress version: 6.1.1
-   Gutenberg version: 14.5.0
-   WooCommerce version: 7.1
-   WooCommerce Blocks version: 8.9.1
-   Site language: English

### Desktop (please complete the following information):

-   OS: Mac and Windows
-   Browser Brave
-   Version v1.45.131, Chromium 107

### PR DESCRIPTION
**Pull request by tarhi-saad**
To avoid limiting the number of edits in WP admin to our Woo Blocks, we need to disable rate limiting.

We disabled rate limiting for users who can edit posts!

Fixes #7738

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

| Before | After |
| ------ | ----- |
|    image    |    image   |

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

Checkout to this branch and:

1. Enable the [rate limiting filter for Store API](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/rate-limiting.md). You can use the [Code Snippets plugin](https://wordpress.org/plugins/code-snippets/) to add the following PHP script:

```PHP
  add_filter( 'woocommerce_store_api_rate_limit_options', function() {
	return [
		'enabled' => true, // enables/disables Rate Limiting. Default: false
		'proxy_support' => false, // enables/disables Proxy support. Default: false
		'limit' => 1, // limit of request per timeframe. Default: 25
		'seconds' => 500, // timeframe in seconds. Default: 10
	];
  } );
```

2. Create a product related page (e.g., Hand-picked Products)
3. Update the block to send some requests to the Store API. In the case of the `Hand-picked Products` block, clicking on `Edit selected products` should do the work:

image

4. Make sure the block is working as expected and you aren't getting any rate-limiting warnings.
5. Open your store in a new window using your browser's incognito mode. Add a product and go to the Cart Block page
6. Increase or decrease the product quantity couple of times. You can see failed rate limited request in the browser dev tools under `Network` tab, or you can reload the page to see the last saved product quantity value. See recording:

https://user-images.githubusercontent.com/14235870/209964938-36018e19-5a0c-414a-8b8b-14683bf7617c.mov

7. Repeat step 6 with a customer role account logged in. You should get the same result from step 6.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin

* [ ] Experimental

### Changelog

> Disable Rate Limiting when editing Blocks in admin

### CODE REVIEW

File: src/StoreApi/Authentication.php [hunk 1]
```diff
@@ -26,6 +26,11 @@ public function check_authentication( $result ) {
 			return $result;
 		}
 
+		// Disable Rate Limiting for logged in users with 'edit posts' capability.
+		if ( current_user_can( 'edit_posts' ) ) {
```
Comment (wavvves) [inline src/StoreApi/Authentication.php]: @tarhi-saad I wonder if at this point, we should move Rate Limiting logic to its own method (it kind of falls outside the naming scope of this one) and avoid the double return of the ternary. Something along the lines of:

```
if ( ! current_user_can( 'edit_posts' ) ) {
    $this->apply_rate_limiting();
}
```

Comment (tarhi-saad) [reply src/StoreApi/Authentication.php]: > I wonder if at this point, we should move Rate Limiting logic to its own method

@wavvves, I've created a separate function for it in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/7934/commits/d608e1c39cb08c9c21ff813b7b83dded9d77eddb). But, for some reason, the rate-limiting wasn't working anymore. So, I reverted the change! Maybe I made a mistake in my refactoring!

**Comment by tarhi-saad** [comment]
Hi @wavvves! 👋 I added some testing instructions to the PR. Would you like to give it a final look? Also, could you check [this comment](https://github.com/woocommerce/woocommerce-blocks/pull/7934#discussion_r1058977031)?

**Comment by wavvves** [comment]
> Hi @wavvves! 👋 I added some testing instructions to the PR. Would you like to give it a final look? Also, could you check [this comment](https://github.com/woocommerce/woocommerce-blocks/pull/7934#discussion_r1058977031)?

Hi Saad, LGTM!
I added a new step to the testing instructions (we should see rate limiting applied to logged-in customers also) and [brought back the refactored](https://github.com/woocommerce/woocommerce-blocks/pull/7934/commits/9427b306aa6409b126b0e0dd5c67425e7ed15074) you did. It was ok, you just missed out on carrying over the returns (my example was too simplistic sorry).

Have another quick test at it, and if tests pass I'll approve it, and it can be merged

**Review by wavvves — APPROVED** [review]

**Comment by wavvves** [comment]
@tarhi-saad I went ahead and approved this since testing passed. Feel free to merge it after you re-checked the testing steps 😎

**Comment by tarhi-saad** [comment]
Thank you for the review and fixes, @wavvves! 🙌 I've run another test, and everything is working as expected! 🎉

**Comment by tobinfekkes** [comment]
Thank you to all involved in getting this update out.

Appreciate it!

**Pull request merged**